sdk/storage/AzureStorageBlob/Source/StorageBlobClient.swift

```diff
@@ -511,7 +512,7 @@ public final class StorageBlobClient: PipelineClient {
         fromContainer container: String,
         toFile destinationUrl: LocalURL,
         withOptions options: DownloadBlobOptions? = nil
-    ) throws -> Transfer? {
+    ) throws -> BlobTransfer? {
```

Comment:
It's not necessary for this, but it makes sense logically. `download(blob:fromContainer:toFile:)` will always start a transfer for a single blob, thus it's appropriate to return a BlobTransfer here. And that way the customer doesn't have to check the type and cast in order to see all of the available properties, all of the properties are much more discoverable this way.

Similarly, when we add a method to do a multi-blob transfer (e.g. something like `downloadAll(fromContainer:toDirectory)`) it should probably return a `MultiBlobTransfer` object.

So in general, I think that we should default to exposing the most specific type we can, since it makes it easier for the customer. And fall back to exposing a generic `Transfer` only when necessary (and the more I think about our API, the fewer places I'm finding it necessary - most places where we might mix together a collection of BlobTransfer and MultiBlobTransfer objects, we probably should instead have separate methods for getting each).

Reply:
I wanted to take another pass after this PR to actually go through and see if I could completely eliminate places where we make the user deal with the `Transfer` protocol. Internally we still use it in quite a few places, but our public API should always know whether a user-facing operation involves a single blob transfer or a multiblob transfer. 

Also, a few thoughts I had as part of the process were:

- We should really encode our states in a state machine, so that it's clear what transitions are valid and how and when they can occur. I started down this path but it was a fairly involved refactor and decided to circle back later.
- We could simplify the TransferDelegate by having only:
  - `transfer(_:didUpdateState:)`
      This is used for any state change including completion and failure. The failed state could contain an associated value with the error.
  - `transfer(_:didUpdateProgress:)`
      Called only when transfers make progress, the UX study seemed to indicate combining this with state was confusing.
- Based on the UX study feedback, should we have a delegate (or could be callbacks) that can be attached to a `Transfer` object itself, which is called only when something changes for that specific transfer? This would be in addition to, not instead of, the client delegate that can be attached to receive updates about all transfers the client creates.

Reply:
I'm okay with returning `BlobTransfer` instead of `Transfer`. The other thoughts should probably be preserved in a separate issue.